hail/src/main/scala/is/hail/expr/ir/analyses/SemanticHash.scala

```diff
@@ -0,0 +1,409 @@
+           _: MatrixFilterEntries |
+           _: MatrixDistinctByRow |
+           _: NDArrayShape |
+           _: NDArraySlice |
+           _: NDArrayReshape |
+           _: NDArrayWrite |
+           _: RelationalLet |
+           _: RNGSplit |
+           _: RNGStateLiteral |
+           _: StreamAgg |
+           _: StreamDrop |
+           _: StreamDropWhile |
+           _: StreamFilter |
+           _: StreamFlatMap |
+           _: StreamFold |
+           _: StreamFold2 |
+           _: StreamFor |
+           _: StreamIota |
+           _: StreamLen |
+           _: StreamMap |
+           _: StreamRange |
+           _: StreamTake |
+           _: StreamTakeWhile |
+           _: TableGetGlobals |
+           _: TableAggregate |
+           _: TableCollect |
+           _: TableCount |
+           _: TableDistinct |
+           _: TableFilter |
+           _: TableMapGlobals |
+           _: TableMapRows |
+           _: TableRename |
+           _: ToArray |
+           _: ToDict |
+           _: ToSet |
+           _: ToStream |
+           _: Trap =>
+        ()
+
+      // Discrete values
+      case _: Void | _: True | _: False =>
+        ()
+
+      // integral constants
+      case I32(x) => buffer ++= Bytes.fromInt(x)
+      case I64(x) => buffer ++= Bytes.fromLong(x)
+      case F32(x) => buffer ++= Bytes.fromFloat(x)
+      case F64(x) => buffer ++= Bytes.fromDouble(x)
+      case NA(typ) => buffer ++= SemanticTypeName(typ)
+      case Str(x) => buffer ++= x.getBytes()
+
+      // In these cases, just return None meaning that two
+      // invocations will never return the same thing.
+      case _ =>
+        log.warn(s"SemanticHash unknown: ${ir.getClass.getName}")
+        return None
+    }
+
+    Some(buffer.result())
+  }
+
+  def getFileHash(fs: FS)(path: String): Array[Byte] =
+    fs.eTag(path) match {
+      case Some(etag) =>
+        etag.getBytes
+      case None =>
+        path.getBytes ++ Bytes.fromLong(fs.fileStatus(path).getModificationTime)
+    }
+
+  def levelOrder(root: BaseIR): Iterator[(BaseIR, Int)] = {
+    val adj: ((BaseIR, Int)) => Iterator[(BaseIR, Int)] =
+      Function.tupled { (ir, trace) =>
+        ir.children
+          .zipWithIndex
+          .map { case (child, k) => (child, extend(trace, Bytes.fromInt(k))) }
+          .iterator
+      }
+
+    TreeTraversal.levelOrder(adj)((root, MurmurHash3.DEFAULT_SEED))
+  }
+
+  object Bytes {
+    def fromInt(x: Int): Array[Byte] =
+      ByteBuffer.allocate(4).putInt(x).array()
+
+    def fromLong(x: Long): Array[Byte] =
+      ByteBuffer.allocate(8).putLong(x).array()
+
+    def fromFloat(x: Float): Array[Byte] =
+      ByteBuffer.allocate(4).putFloat(x).array()
+
+    def fromDouble(x: Double): Array[Byte] =
+      ByteBuffer.allocate(8).putDouble(x).array()
+
+    def fromClass(clz: Class[_]): Array[Byte] =
+      fromInt(clz.hashCode())
+  }
+
+}
+
+case object SemanticTypeName {
+  def apply(t: Type): Array[Byte] = {
```

Comment:
This looks like it does about the same thing as `BaseType.parsableString`. Unless there's a good reason, should probably just use that.

Reply:
The compiler inserts fieldnames that are not normalized

Reply:
Ah, yup, forgot about field names.